### PR TITLE
[run-clang-tidy.py] Refactor, add progress indicator, add type hints

### PR DESCRIPTION
**Pull request by nicovank**
[There is work](https://discourse.llvm.org/t/rfc-upgrading-llvms-minimum-required-python-version/67571) to make Python 3.8 the minimum Python version for LLVM.

I edited this script because I wanted some indicator of progress while going through files.
It now outputs `[XX/YYY]` with the number of processed and total files after each completion.

The current version of this script is compatible downto Python 3.6 (this is PyYAML's minimum version).
It would probably work with older Python 3 versions with an older PyYAML or when YAML is disabled.

With the updates here, it is compatible downto Python 3.7. Python 3.7 was released June 2018.

https://github.com/llvm/llvm-project/pull/89302 is also touching this file, I don't mind rebasing on top of that work if needed.

### Summary

 -  Add type annotations.
 -  Replace `threading` + `queue` with `asyncio`.
 -  **Add indicator of processed files over total files**. This is what I set out to do initially.
 -  Only print the filename after completion, not the entire Clang-Tidy invocation command. I find this neater but the behavior can easily be restored.

### MyPy

```
% python3 -m mypy --strict clang-tools-extra/clang-tidy/tool/run-clang-tidy.py
Success: no issues found in 1 source file
```

### Performance

Just to make sure the `asyncio` change didn't introduce a regression, I ran the previous version and this version 5 times with the following command.

```
time python3 ~/llvm-project/clang-tools-extra/clang-tidy/tool/run-clang-tidy.py \
    -clang-tidy-binary="~/llvm-project/build/bin/clang-tidy" \
    -clang-apply-replacements-binary="~/llvm-project/build/bin/clang-apply-replacements" \
    -checks="-*,modernize-use-starts-ends-with" \
    -source-filter ".*clang-tools-extra/clang-tidy.*" \
    -fix -format
```

Runtimes are identical, no performance regression detected on my setup.

| Version | Time (s)   |
|---------|------------|
| Old     | 92.0 ± 1.4 |
| New     | 92.2 ± 1.5 |

### CODE REVIEW

**Comment by nicovank** [comment]
@PiotrZSL @carlosgalvezp Any thoughts on this change?

**Comment by PiotrZSL** [comment]
I'm more or less fine with this change, but didn't had time to test it yet.
Does it work with python 3.6 ?

**Comment by nicovank** [comment]
> Does it work with Python 3.6?

Actually, it can with one single change. `asyncio,run` was introduced in 3.7, and can be replaced. I'll just add this change and a comment. It worked on my 3.6 setup. Let me know if you run into any issues.

File: clang-tools-extra/clang-tidy/tool/run-clang-tidy.py
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/tool/run-clang-tidy.py:524]: Make sure to add flush to both buffers after printing.

**Review by PiotrZSL — COMMENTED** [review]
One issue, looks like only stdout is buffered, and when clang-tidy output:

"26035 warnings generated.
Suppressed 26030 warnings (26030 in non-user code).
Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.

26005 warnings generated.
Suppressed 26002 warnings (26002 in non-user code).
Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.

25999 warnings generated.
Suppressed 25996 warnings (25996 in non-user code).
Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.
"

it's just displayed without point to an file that produced this output.
This is what I see when redirecting output to file with "&> output.txt".
Looks like without redirection to file works fine.

Second issue, is that it takes a while before first output is being shown to user.
Would be nice if we would show something like "Running clang-tidy for XYZ files out of ZXY in compilation database", or something like this.

You may consider adding also an timer how long command took for every file.
Could be usefull

**Comment by nicovank** [comment]
Reproduced the buffering issue on my setup, but flushing didn't fix the problem.
I could also reproduce with the current version of this script, which does also flush after writing.
Python defaults input/output streams to buffered mode when it detects a pipe, and I don't think flushing overrides that setting.
One way to force unbuffering I know of is to run `python3 -u` when initially invoking the script, but that's not great here.

<details>

<summary>Reproduction with current script version.</summary>

```
% python3 run-clang-tidy.py -checks="-*,performance-*" -source-filter ".*clang-tools-extra/clang-query.*" &>output

/home/nvankempen_umass_edu/work/build/bin/clang-tidy -checks=-*,performance-* -p=/home/nvankempen_umass_edu/work/llvm-project /home/nvankempen_umass_edu/work/llvm-project/clang-tools-extra/clang-query/QueryParser.cpp
/home/nvankempen_umass_edu/work/llvm-project/clang-tools-extra/clang-query/QueryParser.cpp:175:6: warning: enum 'ParsedQueryKind' uses a larger base type ('unsigned int', size: 4 bytes) than necessary for its value set, consider using 'std::uint8_t' (1 byte) as the base type to reduce its size [performance-enum-size]
  175 | enum ParsedQueryKind {
      |      ^
/home/nvankempen_umass_edu/work/llvm-project/clang-tools-extra/clang-query/QueryParser.cpp:189:6: warning: enum 'ParsedQueryVariable' uses a larger base type ('unsigned int', size: 4 bytes) than necessary for its value set, consider using 'std::uint8_t' (1 byte) as the base type to reduce its size [performance-enum-size]
  189 | enum ParsedQueryVariable {
      |      ^
/home/nvankempen_umass_edu/work/build/bin/clang-tidy -checks=-*,performance-* -p=/home/nvankempen_umass_edu/work/llvm-project /home/nvankempen_umass_edu/work/llvm-project/clang-tools-extra/clang-query/tool/ClangQuery.cpp
/home/nvankempen_umass_edu/work/build/bin/clang-tidy -checks=-*,performance-* -p=/home/nvankempen_umass_edu/work/llvm-project /home/nvankempen_umass_edu/work/llvm-project/clang-tools-extra/clang-query/Query.cpp
1363 warnings generated.
Suppressed 1361 warnings (1361 in non-user code).
Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.
1595 warnings generated.
Suppressed 1595 warnings (1595 in non-user code).
Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.
1633 warnings generated.
Suppressed 1633 warnings (1633 in non-user code).
Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.
```

</details>

I think the way to go here is just to print everything to stdout, stderr can be reserved to issues within the script itself.
**→ This is the behavior in the current version of this PR**.

Added an initial message with the number of files to process over the number of files in compilation database.

Added a runtime number for each file.
Thought about making a `ClangTidyResult` class to wrap the `run_tidy` return, but it's only used once and would make the deconstructing harder, so it's fine.

**Comment by nicovank** [comment]
Rebased on top of cc54129b983799e1aaea77aa0ff3040dc30cbc8c.

Ping @PiotrZSL, what to do here? I can keep the current flush and separate stdout/stderr behavior in this PR, it can be changed later.
But I don't think flushing fixes that issue of mis-ordered output lines, IMO either 1. use the same stream to force order 2. let user unbuffer with the `python3 -u` flag.

**Comment by PiotrZSL** [comment]
@nicovank I think that code is fine, will re-check it today. We could always think about something else later.

**Comment by PiotrZSL** [comment]
@nicovank  Overall looks fine from functionality point of view.

Problem A)
Getting bunch of
```
task: <Task pending name='Task-4805' coro=<run_with_semaphore() done, defined at run-clang-tidy.py:211> wait_for=<Future pending cb=[Task.task_wakeup()]> cb=[as_completed.<locals>._on_completion() at /usr/lib/python3.12/asyncio/tasks.py:618]>
Task was destroyed but it is pending!
```
When CTR+C a script. More graceful shutdown would be welcome.

Problem B)
Previously script printed clang-tidy command line, now just progress + filename, for me it's fine but some users could find this as an degradation, as previously you could simply copy command line output and just run it manually.

My advice:
- By default print like in original script (command line), but add --progress switch that will change it to current behavior.
- Add release notes entry about updates in script that impact end user.

File: clang-tools-extra/clang-tidy/tool/run-clang-tidy.py
Comment (5chmidti) [inline clang-tools-extra/clang-tidy/tool/run-clang-tidy.py:577]: Just asking: Being unfamiliar with `asyncio`, why do we need this part and an `async main`, instead of calling plain `main`?

**Review by 5chmidti — COMMENTED** [review]
Thanks for the cleanup, it looks overall good (w.r.t `asyncio`: I only know about `asyncio` what I read in this pr).

> Only print the filename after completion, not the entire Clang-Tidy invocation command. I find this neater but the behavior can easily be restored.

> By default print like in original script (command line), but add --progress switch that will change it to current behavior.

I'd prefer the original behavior as well, but printing only the filename can be added as a flag. Although, I'm not sure `--progress` is a good name for something that only changes if the command or file is printed, because the progress indicator should be emitted either way.

> Added an initial message with the number of files to process over the number of files in compilation database.

Maybe that was removed? Either way, I think it's +- not necessary, but I'm not against having it.

> More graceful shutdown would be welcome.

+1
IMO, that needs to be done before merging.

File: clang-tools-extra/clang-tidy/tool/run-clang-tidy.py
Comment (nicovank) [reply clang-tools-extra/clang-tidy/tool/run-clang-tidy.py:577]: If you call main or any other `async` function without special handling you get some
```
RuntimeWarning: coroutine 'main' was never awaited
```

Since `await` can only be used in `async` functions, it's common to use [`asyncio.run`](https://docs.python.org/3/library/asyncio-runner.html#asyncio.run) and wrap `main`. That is Python 3.7 and above, we get compatibility with 3.6 by using those 3 lines instead.

**Comment by nicovank** [comment]
-  Minor change to progress indicator to stay aligned.
 -  Go back to printing command line. I'm fine with leaving this as-is.
 -  I think I pruned some changes by accident on rebase, thanks @5chmidti! Restored runtime number and initial message.

--

> More graceful shutdown would be welcome.

@PiotrZSL I can't reproduce on my setup (MacOS/3.12). What OS/Python version does this occur on? Will look into better shutdown.

File: clang-tools-extra/clang-tidy/tool/run-clang-tidy.py
Comment (5chmidti) [reply clang-tools-extra/clang-tidy/tool/run-clang-tidy.py:616]: Sure, but why does `main` need to be `async` at all?

I.e.

```python
async def main():
  pass

if __name__ == "__main__":
    # FIXME Python 3.7: This can be simplified by asyncio.run(main()).
    loop = asyncio.new_event_loop()
    loop.run_until_complete(main())
    loop.close()
```

vs.

```python
def main():
  pass

if __name__ == "__main__":
    main()
```

**Review by 5chmidti — COMMENTED** [review]
The info about the execution time of clang-tidy is a nice improvement as well.

---

> What OS/Python version does this occur on?

|OS|Python|Noisy Shutdown|
|-|-|-|
|Manjaro|3.11.9|Yes|
|NixOS|3.9.19|No|
|NixOS|3.10.14|No|
|NixOS|3.11.9|No|
|NixOS|3.12.3|Yes|
|NixOS|3.13.0b2|Yes|

**Comment by PiotrZSL** [comment]
Note: 4 days to branch-out

File: clang-tools-extra/clang-tidy/tool/run-clang-tidy.py
Comment (nicovank) [reply clang-tools-extra/clang-tidy/tool/run-clang-tidy.py:616]: Because it uses one `await` keyword: `result = await coro`, which can only be present in an `async` function. Here it basically awaits the first available result from `tasks`.

**Comment by nicovank** [comment]
Thanks for the ping, I had a version for 3.7+ working, and it seems hard to retrofit to 3.6. I think having the 3.7 version is fine. It's on my home machine, will update this PR later today 👍

**Comment by nicovank** [comment]
Update: shutdown is now graceful on my machine. Please test on others if possible. 3.7+ is now required, which I think is fine given we want to make 3.8 the minimum across LLVM.

Rebased on top of the `allow-no-checks` commit.

**Review by 5chmidti — APPROVED** [review]
Stopping the execution is now graceful on my machine as well, thanks.

**Comment by nicovank** [comment]
Just double checked and tested again, seems good to me.
Please merge on my behalf, I do not have commit access. I'll keep an eye out if anything breaks.

**Review by PiotrZSL — APPROVED** [review]

**Pull request merged**
